{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8c753145-1a59-480d-abb2-60f2fa3f416f",
   "metadata": {},
   "source": [
    "# Inputs\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2b6aeb01",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np, zarr, nrrd\n",
    "# alignment functions\n",
    "from bigstream.align import alignment_pipeline\n",
    "from bigstream.transform import apply_transform\n",
    "import tifffile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f9e34109-0b59-441f-bb7a-2f75c90d30eb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "p = '/u/home/f/f7xiesnm/project-zipursky/easifish/lt186/outputs/'\n",
    "outdir = '/u/home/f/f7xiesnm/project-zipursky/easifish/results/test_bigstream2'\n",
    "!mkdir -p $outdir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5d71f6f4-5204-4d2d-8790-249b7b7b1fa3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.21  0.115 0.115] [0.21  0.115 0.115]\n",
      "[0.21 0.23 0.23] [0.21 0.23 0.23]\n",
      "[0.84 0.92 0.92] [0.84 0.92 0.92]\n",
      "[1.89 1.84 1.84] [1.89 1.84 1.84]\n",
      "(2593, 7309, 9049) (2296, 7336, 9091)\n",
      "(648, 913, 1131) (574, 917, 1136)\n",
      "(288, 456, 565) (255, 458, 568)\n"
     ]
    }
   ],
   "source": [
    "fix_path = p + '/r1/stitching/export.n5'\n",
    "mov_path = p + '/r7/stitching/export.n5'\n",
    "exp_factor = 2  # replace this with the known expansion factor for your sample to use pre-expansion units\n",
    "\n",
    "s1_scale = [1,2,2] # z,y,x\n",
    "s3_scale = [4,8,8]\n",
    "s4_scale = [9,16,16]\n",
    "\n",
    "# load fix data and spacing\n",
    "fix_zarr = zarr.open(store=zarr.N5Store(fix_path), mode='r')\n",
    "fix_meta = fix_zarr.attrs.asdict()\n",
    "fix_spacing = np.array(fix_meta['pixelResolution']['dimensions'][::-1]) / exp_factor\n",
    "fix_spacing_s1 = fix_spacing * s1_scale \n",
    "fix_spacing_s3 = fix_spacing * s3_scale \n",
    "fix_spacing_s4 = fix_spacing * s4_scale \n",
    "\n",
    "# load mov data and spacing\n",
    "mov_zarr = zarr.open(store=zarr.N5Store(mov_path), mode='r')\n",
    "mov_meta = mov_zarr.attrs.asdict()\n",
    "mov_spacing = np.array(mov_meta['pixelResolution']['dimensions'][::-1]) / exp_factor\n",
    "mov_spacing_s1 = mov_spacing * s1_scale\n",
    "mov_spacing_s3 = mov_spacing * s3_scale \n",
    "mov_spacing_s4 = mov_spacing * s4_scale \n",
    "\n",
    "# print spacings\n",
    "print(fix_spacing, mov_spacing)\n",
    "print(fix_spacing_s1, mov_spacing_s1)\n",
    "print(fix_spacing_s3, mov_spacing_s3)\n",
    "print(fix_spacing_s4, mov_spacing_s4)\n",
    "\n",
    "# print shapes\n",
    "print(fix_zarr['/c3/s0'].shape, mov_zarr['/c3/s0'].shape)\n",
    "print(fix_zarr['/c3/s3'].shape, mov_zarr['/c3/s3'].shape)\n",
    "print(fix_zarr['/c3/s4'].shape, mov_zarr['/c3/s4'].shape)\n",
    "\n",
    "# write some channels\n",
    "nrrd.write(outdir+'/fix.nrrd', fix_zarr['/c3/s4'][...].transpose(2,1,0), compression_level=2)\n",
    "nrrd.write(outdir+'/mov.nrrd', mov_zarr['/c3/s4'][...].transpose(2,1,0), compression_level=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "282cb995-1ffa-45ff-8d12-49a3cf6cab81",
   "metadata": {},
   "source": [
    "# Alignment\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "102718cb-4dee-4a74-9056-6496692caed8",
   "metadata": {},
   "source": [
    "### global"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "54f5483b-6197-4441-89e9-d944290ae7d6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fewer than 50 spots found in fixed image, returning default\n",
      "LEVEL:  0  ITERATION:  0  METRIC:  -0.0060492718906934055\n",
      "LEVEL:  0  ITERATION:  1  METRIC:  -0.006090899593973053\n",
      "LEVEL:  0  ITERATION:  2  METRIC:  -0.006073962860635501\n",
      "LEVEL:  0  ITERATION:  3  METRIC:  -0.005988613324086965\n",
      "LEVEL:  0  ITERATION:  4  METRIC:  -0.006176481572129269\n",
      "LEVEL:  0  ITERATION:  5  METRIC:  -0.006381889463013508\n",
      "LEVEL:  0  ITERATION:  6  METRIC:  -0.00652736332496646\n",
      "LEVEL:  0  ITERATION:  7  METRIC:  -0.006604594577178613\n",
      "LEVEL:  0  ITERATION:  8  METRIC:  -0.006613979217230575\n",
      "LEVEL:  0  ITERATION:  9  METRIC:  -0.006566979052232072\n",
      "LEVEL:  0  ITERATION:  10  METRIC:  -0.006466691532997087\n",
      "LEVEL:  0  ITERATION:  11  METRIC:  -0.006559428235691139\n",
      "LEVEL:  0  ITERATION:  12  METRIC:  -0.006747658593064638\n",
      "LEVEL:  0  ITERATION:  13  METRIC:  -0.006866183983007629\n",
      "LEVEL:  0  ITERATION:  14  METRIC:  -0.006912984051493983\n",
      "LEVEL:  0  ITERATION:  15  METRIC:  -0.006903676642438746\n",
      "LEVEL:  0  ITERATION:  16  METRIC:  -0.006841874241334527\n",
      "LEVEL:  0  ITERATION:  17  METRIC:  -0.006723902801939866\n",
      "LEVEL:  0  ITERATION:  18  METRIC:  -0.006603228324413273\n",
      "LEVEL:  0  ITERATION:  19  METRIC:  -0.006778052549899519\n",
      "LEVEL:  0  ITERATION:  20  METRIC:  -0.006912468891542923\n",
      "LEVEL:  0  ITERATION:  21  METRIC:  -0.006993132136930052\n",
      "LEVEL:  0  ITERATION:  22  METRIC:  -0.007009796604979791\n",
      "LEVEL:  0  ITERATION:  23  METRIC:  -0.006956374870633647\n",
      "LEVEL:  0  ITERATION:  24  METRIC:  -0.00685425076820295\n",
      "LEVEL:  0  ITERATION:  25  METRIC:  -0.0067113083907271364\n",
      "LEVEL:  0  ITERATION:  26  METRIC:  -0.006540926620676925\n",
      "LEVEL:  0  ITERATION:  27  METRIC:  -0.00638709131224204\n",
      "Optimization failed to improve metric\n",
      "METRIC VALUES initial: -0.00019407972422497185 final: -0.00017627787755992132\n",
      "Returning default\n",
      "CPU times: user 10min 26s, sys: 47.8 s, total: 11min 14s\n",
      "Wall time: 5min 10s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# alignment functions\n",
    "from bigstream.align import alignment_pipeline\n",
    "from bigstream.transform import apply_transform\n",
    "\n",
    "# get global alignment channels\n",
    "fix = fix_zarr['/c3/s4'][...]\n",
    "mov = mov_zarr['/c3/s4'][...] #[:, ::-1, ::-1]  # moving image data is 180 degrees rotates about z axis\n",
    "\n",
    "# create some masks to focus alignment only on planes below the large artifact at the top\n",
    "fix_mask = np.ones(fix.shape, dtype=np.uint8)\n",
    "fix_mask[:65] = 0\n",
    "mov_mask = np.ones(mov.shape, dtype=np.uint8)\n",
    "mov_mask[:45] = 0\n",
    "\n",
    "# define alignment steps\n",
    "ransac_kwargs = {\n",
    "    'blob_sizes':[2, 8],\n",
    "    'cc_radius':12,\n",
    "    'match_threshold':0.4,\n",
    "    'nspots':10000,\n",
    "}\n",
    "\n",
    "affine_kwargs = {\n",
    "    'alignment_spacing':4.0,\n",
    "    'shrink_factors':(1,),\n",
    "    'smooth_sigmas':(4.,),\n",
    "    'optimizer_args':{\n",
    "        'learningRate':0.25,\n",
    "        'minStep':0.,\n",
    "        'numberOfIterations':400,\n",
    "    },\n",
    "    'metric':'C',\n",
    "}\n",
    "\n",
    "steps = [('ransac', ransac_kwargs,),\n",
    "         ('affine', affine_kwargs,),]\n",
    "\n",
    "# align\n",
    "affine = alignment_pipeline(\n",
    "    fix, mov,\n",
    "    fix_spacing_s4,\n",
    "    mov_spacing_s4,\n",
    "    steps,\n",
    "    fix_mask=fix_mask,\n",
    "    mov_mask=mov_mask,\n",
    ")\n",
    "\n",
    "# apply affine only\n",
    "affine_aligned = apply_transform(\n",
    "    fix, mov,\n",
    "    fix_spacing_s4, mov_spacing_s4,\n",
    "    transform_list=[affine,],\n",
    ")\n",
    "\n",
    "# write results\n",
    "np.savetxt(outdir+'/affine.mat', affine)\n",
    "nrrd.write(outdir+'/affine.nrrd', affine_aligned.transpose(2,1,0), compression_level=2)\n",
    "tifffile.imwrite(outdir+'/affine.tiff', affine_aligned.transpose([2,1,0]))\n",
    "\n",
    "# load precomputed results\n",
    "affine = np.loadtxt(outdir+'/affine.mat')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "04ab7eb9-5e13-4475-a999-7499632e85bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fewer than 50 spots found in fixed image, returning default\n",
      "Optimization failed to improve metric\n",
      "METRIC VALUES initial: -7.59456015081018e-05 final: -7.59456015081018e-05\n",
      "Returning default\n",
      "CPU times: user 4min 23s, sys: 39 s, total: 5min 3s\n",
      "Wall time: 4min 28s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# alignment functions\n",
    "from bigstream.align import alignment_pipeline\n",
    "from bigstream.transform import apply_transform\n",
    "\n",
    "# get global alignment channels\n",
    "fix = fix_zarr['/c3/s4'][...]\n",
    "mov = mov_zarr['/c3/s4'][...][:, ::-1, ::-1]  # moving image data is 180 degrees rotates about z axis\n",
    "\n",
    "# create some masks to focus alignment only on planes below the large artifact at the top\n",
    "fix_mask = np.ones(fix.shape, dtype=np.uint8)\n",
    "fix_mask[:65] = 0\n",
    "mov_mask = np.ones(mov.shape, dtype=np.uint8)\n",
    "mov_mask[:45] = 0\n",
    "\n",
    "# define alignment steps\n",
    "ransac_kwargs = {\n",
    "    'blob_sizes':[2, 8],\n",
    "    'cc_radius':12,\n",
    "    'match_threshold':0.4,\n",
    "    'nspots':10000,\n",
    "}\n",
    "\n",
    "affine_kwargs = {\n",
    "    'alignment_spacing':4.0,\n",
    "    'shrink_factors':(1,),\n",
    "    'smooth_sigmas':(4.,),\n",
    "    'optimizer_args':{\n",
    "        'learningRate':0.25,\n",
    "        'minStep':0.,\n",
    "        'numberOfIterations':400,\n",
    "    },\n",
    "    'metric':'C',\n",
    "}\n",
    "\n",
    "steps = [('ransac', ransac_kwargs,),\n",
    "         ('affine', affine_kwargs,),]\n",
    "\n",
    "# align\n",
    "affine = alignment_pipeline(\n",
    "    fix, mov,\n",
    "    fix_spacing_s4,\n",
    "    mov_spacing_s4,\n",
    "    steps,\n",
    "    fix_mask=fix_mask,\n",
    "    mov_mask=mov_mask,\n",
    ")\n",
    "\n",
    "# apply affine only\n",
    "affine_aligned = apply_transform(\n",
    "    fix, mov,\n",
    "    fix_spacing_s4, mov_spacing_s4,\n",
    "    transform_list=[affine,],\n",
    ")\n",
    "\n",
    "# write results\n",
    "np.savetxt(outdir+'/affine2.mat', affine)\n",
    "nrrd.write(outdir+'/affine2.nrrd', affine_aligned.transpose(2,1,0), compression_level=2)\n",
    "tifffile.imwrite(outdir+'/affine2.tiff', affine_aligned.transpose([2,1,0]))\n",
    "\n",
    "# load precomputed results\n",
    "affine = np.loadtxt(outdir+'/affine2.mat')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a11435c0-f647-409f-a47d-b2004bace0e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d140507f-5ba0-4c00-a2b4-baeac6057ae6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dcf3a8ce-5693-4b1b-83b4-8ca1169b85fc",
   "metadata": {},
   "source": [
    "### local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d3d8b2f6-8af4-420a-9c3d-ba09fdb0343f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'bigstream.piecewise_align' from '/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/bigstream/piecewise_align.py'>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from bigstream.piecewise_align import distributed_piecewise_alignment_pipeline\n",
    "from bigstream.transform import apply_transform\n",
    "\n",
    "import importlib\n",
    "import bigstream\n",
    "import bigstream.piecewise_align\n",
    "importlib.reload(bigstream.piecewise_align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "25bf63b9-1732-4b54-b8f2-6ba05b6f0fa8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-24 15:52:43,613 - distributed.diskutils - INFO - Found stale lock file and directory '/tmp/dask-worker-space/worker-ekofv7z0', purging\n",
      "2023-07-24 15:52:43,613 - distributed.diskutils - INFO - Found stale lock file and directory '/tmp/dask-worker-space/worker-vo0m7xn4', purging\n",
      "2023-07-24 15:52:43,613 - distributed.diskutils - INFO - Found stale lock file and directory '/tmp/dask-worker-space/worker-6tw9hbpi', purging\n",
      "2023-07-24 15:52:43,613 - distributed.diskutils - INFO - Found stale lock file and directory '/tmp/dask-worker-space/worker-6u_olvwk', purging\n",
      "2023-07-24 15:52:43,619 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "2023-07-24 15:52:43,619 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "2023-07-24 15:52:43,619 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "2023-07-24 15:52:43,619 - distributed.nanny - ERROR - Failed to initialize worker\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "2023-07-24 15:52:43,640 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 443, in instantiate\n",
      "    result = await self.process.start()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 720, in start\n",
      "    msg = await self._wait_until_connected(uid)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 868, in _wait_until_connected\n",
      "    raise msg[\"exception\"]\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "2023-07-24 15:52:43,643 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 443, in instantiate\n",
      "    result = await self.process.start()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 720, in start\n",
      "    msg = await self._wait_until_connected(uid)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 868, in _wait_until_connected\n",
      "    raise msg[\"exception\"]\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "2023-07-24 15:52:43,645 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 443, in instantiate\n",
      "    result = await self.process.start()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 720, in start\n",
      "    msg = await self._wait_until_connected(uid)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 868, in _wait_until_connected\n",
      "    raise msg[\"exception\"]\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "2023-07-24 15:52:43,652 - distributed.nanny - ERROR - Failed to start process\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 443, in instantiate\n",
      "    result = await self.process.start()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 720, in start\n",
      "    msg = await self._wait_until_connected(uid)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 868, in _wait_until_connected\n",
      "    raise msg[\"exception\"]\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-207' coro=<_wrap_awaitable() done, defined at /u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py:124> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py\", line 537, in start\n",
      "    await wait_for(self.start_unsafe(), timeout=timeout)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/utils.py\", line 1878, in wait_for\n",
      "    return await asyncio.wait_for(fut, timeout)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/asyncio/tasks.py\", line 442, in wait_for\n",
      "    return await fut\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 369, in start_unsafe\n",
      "    response = await self.instantiate()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 443, in instantiate\n",
      "    result = await self.process.start()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 720, in start\n",
      "    msg = await self._wait_until_connected(uid)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 868, in _wait_until_connected\n",
      "    raise msg[\"exception\"]\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py\", line 125, in _wrap_awaitable\n",
      "    return await aw\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py\", line 545, in start\n",
      "    raise RuntimeError(f\"{type(self).__name__} failed to start.\") from exc\n",
      "RuntimeError: Nanny failed to start.\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-206' coro=<_wrap_awaitable() done, defined at /u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py:124> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py\", line 537, in start\n",
      "    await wait_for(self.start_unsafe(), timeout=timeout)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/utils.py\", line 1878, in wait_for\n",
      "    return await asyncio.wait_for(fut, timeout)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/asyncio/tasks.py\", line 442, in wait_for\n",
      "    return await fut\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 369, in start_unsafe\n",
      "    response = await self.instantiate()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 443, in instantiate\n",
      "    result = await self.process.start()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 720, in start\n",
      "    msg = await self._wait_until_connected(uid)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 868, in _wait_until_connected\n",
      "    raise msg[\"exception\"]\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py\", line 125, in _wrap_awaitable\n",
      "    return await aw\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py\", line 545, in start\n",
      "    raise RuntimeError(f\"{type(self).__name__} failed to start.\") from exc\n",
      "RuntimeError: Nanny failed to start.\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-205' coro=<_wrap_awaitable() done, defined at /u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py:124> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py\", line 537, in start\n",
      "    await wait_for(self.start_unsafe(), timeout=timeout)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/utils.py\", line 1878, in wait_for\n",
      "    return await asyncio.wait_for(fut, timeout)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/asyncio/tasks.py\", line 442, in wait_for\n",
      "    return await fut\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 369, in start_unsafe\n",
      "    response = await self.instantiate()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 443, in instantiate\n",
      "    result = await self.process.start()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 720, in start\n",
      "    msg = await self._wait_until_connected(uid)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 868, in _wait_until_connected\n",
      "    raise msg[\"exception\"]\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py\", line 125, in _wrap_awaitable\n",
      "    return await aw\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py\", line 545, in start\n",
      "    raise RuntimeError(f\"{type(self).__name__} failed to start.\") from exc\n",
      "RuntimeError: Nanny failed to start.\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-208' coro=<_wrap_awaitable() done, defined at /u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py:124> exception=RuntimeError('Nanny failed to start.')>\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py\", line 537, in start\n",
      "    await wait_for(self.start_unsafe(), timeout=timeout)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/utils.py\", line 1878, in wait_for\n",
      "    return await asyncio.wait_for(fut, timeout)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/asyncio/tasks.py\", line 442, in wait_for\n",
      "    return await fut\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 369, in start_unsafe\n",
      "    response = await self.instantiate()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 443, in instantiate\n",
      "    result = await self.process.start()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 720, in start\n",
      "    msg = await self._wait_until_connected(uid)\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 868, in _wait_until_connected\n",
      "    raise msg[\"exception\"]\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py\", line 919, in run\n",
      "    worker = worker_factory()\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py\", line 767, in __init__\n",
      "    ServerNode.__init__(\n",
      "TypeError: __init__() got an unexpected keyword argument 'project'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py\", line 125, in _wrap_awaitable\n",
      "    return await aw\n",
      "  File \"/u/home/f/f7xiesnm/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py\", line 545, in start\n",
      "    raise RuntimeError(f\"{type(self).__name__} failed to start.\") from exc\n",
      "RuntimeError: Nanny failed to start.\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "__init__() got an unexpected keyword argument 'project'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "File \u001b[0;32m<timed exec>:39\u001b[0m\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/ClusterWrap/decorator.py:21\u001b[0m, in \u001b[0;36mcluster.<locals>.create_or_pass_cluster\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m cluster \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m     18\u001b[0m \n\u001b[1;32m     19\u001b[0m     \u001b[38;5;66;03m# get arguments and run function in cluster context\u001b[39;00m\n\u001b[1;32m     20\u001b[0m     x \u001b[38;5;241m=\u001b[39m kwargs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcluster_kwargs\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcluster_kwargs\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m kwargs \u001b[38;5;28;01melse\u001b[39;00m {}\n\u001b[0;32m---> 21\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mcluster_constructor\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m cluster:\n\u001b[1;32m     22\u001b[0m         kwargs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcluster\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m cluster\n\u001b[1;32m     23\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/ClusterWrap/clusters.py:200\u001b[0m, in \u001b[0;36mlocal_cluster.__init__\u001b[0;34m(self, config, memory_limit, **kwargs)\u001b[0m\n\u001b[1;32m    197\u001b[0m     kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmemory_limit\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m memory_limit\n\u001b[1;32m    199\u001b[0m \u001b[38;5;66;03m# set up cluster, connect scheduler/client\u001b[39;00m\n\u001b[0;32m--> 200\u001b[0m cluster \u001b[38;5;241m=\u001b[39m \u001b[43mLocalCluster\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    201\u001b[0m client \u001b[38;5;241m=\u001b[39m Client(cluster)\n\u001b[1;32m    202\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mset_cluster(cluster)\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/local.py:253\u001b[0m, in \u001b[0;36mLocalCluster.__init__\u001b[0;34m(self, name, n_workers, threads_per_worker, processes, loop, start, host, ip, scheduler_port, silence_logs, dashboard_address, worker_dashboard_address, diagnostics_port, services, worker_services, service_kwargs, asynchronous, security, protocol, blocked_handlers, interface, worker_class, scheduler_kwargs, scheduler_sync_interval, **worker_kwargs)\u001b[0m\n\u001b[1;32m    250\u001b[0m worker \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcls\u001b[39m\u001b[38;5;124m\"\u001b[39m: worker_class, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moptions\u001b[39m\u001b[38;5;124m\"\u001b[39m: worker_kwargs}\n\u001b[1;32m    251\u001b[0m workers \u001b[38;5;241m=\u001b[39m {i: worker \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(n_workers)}\n\u001b[0;32m--> 253\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    254\u001b[0m \u001b[43m    \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    255\u001b[0m \u001b[43m    \u001b[49m\u001b[43mscheduler\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mscheduler\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    256\u001b[0m \u001b[43m    \u001b[49m\u001b[43mworkers\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mworkers\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    257\u001b[0m \u001b[43m    \u001b[49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    258\u001b[0m \u001b[43m    \u001b[49m\u001b[43mloop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mloop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    259\u001b[0m \u001b[43m    \u001b[49m\u001b[43masynchronous\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43masynchronous\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    260\u001b[0m \u001b[43m    \u001b[49m\u001b[43msilence_logs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msilence_logs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    261\u001b[0m \u001b[43m    \u001b[49m\u001b[43msecurity\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msecurity\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    262\u001b[0m \u001b[43m    \u001b[49m\u001b[43mscheduler_sync_interval\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mscheduler_sync_interval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    263\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py:293\u001b[0m, in \u001b[0;36mSpecCluster.__init__\u001b[0;34m(self, workers, scheduler, worker, asynchronous, loop, security, silence_logs, name, shutdown_on_close, scheduler_sync_interval)\u001b[0m\n\u001b[1;32m    291\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msync(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_start)\n\u001b[1;32m    292\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 293\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msync\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_correct_state\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    294\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m    295\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msync(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose)\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/utils.py:351\u001b[0m, in \u001b[0;36mSyncMethodMixin.sync\u001b[0;34m(self, func, asynchronous, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    349\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m future\n\u001b[1;32m    350\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 351\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msync\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    352\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallback_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback_timeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\n\u001b[1;32m    353\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/utils.py:418\u001b[0m, in \u001b[0;36msync\u001b[0;34m(loop, func, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    416\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m error:\n\u001b[1;32m    417\u001b[0m     typ, exc, tb \u001b[38;5;241m=\u001b[39m error\n\u001b[0;32m--> 418\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc\u001b[38;5;241m.\u001b[39mwith_traceback(tb)\n\u001b[1;32m    419\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    420\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/utils.py:391\u001b[0m, in \u001b[0;36msync.<locals>.f\u001b[0;34m()\u001b[0m\n\u001b[1;32m    389\u001b[0m         future \u001b[38;5;241m=\u001b[39m wait_for(future, callback_timeout)\n\u001b[1;32m    390\u001b[0m     future \u001b[38;5;241m=\u001b[39m asyncio\u001b[38;5;241m.\u001b[39mensure_future(future)\n\u001b[0;32m--> 391\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01myield\u001b[39;00m future\n\u001b[1;32m    392\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m    393\u001b[0m     error \u001b[38;5;241m=\u001b[39m sys\u001b[38;5;241m.\u001b[39mexc_info()\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/tornado/gen.py:769\u001b[0m, in \u001b[0;36mRunner.run\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    766\u001b[0m exc_info \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    768\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 769\u001b[0m     value \u001b[38;5;241m=\u001b[39m \u001b[43mfuture\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mresult\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    770\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m    771\u001b[0m     exc_info \u001b[38;5;241m=\u001b[39m sys\u001b[38;5;241m.\u001b[39mexc_info()\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/deploy/spec.py:387\u001b[0m, in \u001b[0;36mSpecCluster._correct_state_internal\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    385\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m w \u001b[38;5;129;01min\u001b[39;00m workers:\n\u001b[1;32m    386\u001b[0m         w\u001b[38;5;241m.\u001b[39m_cluster \u001b[38;5;241m=\u001b[39m weakref\u001b[38;5;241m.\u001b[39mref(\u001b[38;5;28mself\u001b[39m)\n\u001b[0;32m--> 387\u001b[0m         \u001b[38;5;28;01mawait\u001b[39;00m w  \u001b[38;5;66;03m# for tornado gen.coroutine support\u001b[39;00m\n\u001b[1;32m    388\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mworkers\u001b[38;5;241m.\u001b[39mupdate(\u001b[38;5;28mdict\u001b[39m(\u001b[38;5;28mzip\u001b[39m(to_open, workers)))\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py:526\u001b[0m, in \u001b[0;36mServer.start\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstatus \u001b[38;5;241m==\u001b[39m Status\u001b[38;5;241m.\u001b[39mfailed:\n\u001b[1;32m    525\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__startup_exc \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 526\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__startup_exc\n\u001b[1;32m    527\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstatus \u001b[38;5;241m!=\u001b[39m Status\u001b[38;5;241m.\u001b[39minit:\n\u001b[1;32m    528\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/core.py:537\u001b[0m, in \u001b[0;36mServer.start\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    534\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__startup_exc \u001b[38;5;241m=\u001b[39m exc\n\u001b[1;32m    536\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 537\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m wait_for(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstart_unsafe(), timeout\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    538\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m asyncio\u001b[38;5;241m.\u001b[39mTimeoutError \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    539\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m _close_on_failure(exc)\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/utils.py:1878\u001b[0m, in \u001b[0;36mwait_for\u001b[0;34m(fut, timeout)\u001b[0m\n\u001b[1;32m   1877\u001b[0m \u001b[38;5;28;01masync\u001b[39;00m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwait_for\u001b[39m(fut: Awaitable[T], timeout: \u001b[38;5;28mfloat\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m T:\n\u001b[0;32m-> 1878\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m asyncio\u001b[38;5;241m.\u001b[39mwait_for(fut, timeout)\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/asyncio/tasks.py:442\u001b[0m, in \u001b[0;36mwait_for\u001b[0;34m(fut, timeout, loop)\u001b[0m\n\u001b[1;32m    437\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe loop argument is deprecated since Python 3.8, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    438\u001b[0m                   \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mand scheduled for removal in Python 3.10.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    439\u001b[0m                   \u001b[38;5;167;01mDeprecationWarning\u001b[39;00m, stacklevel\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m)\n\u001b[1;32m    441\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 442\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m fut\n\u001b[1;32m    444\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    445\u001b[0m     fut \u001b[38;5;241m=\u001b[39m ensure_future(fut, loop\u001b[38;5;241m=\u001b[39mloop)\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py:369\u001b[0m, in \u001b[0;36mNanny.start_unsafe\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mplugin_add(plugin\u001b[38;5;241m=\u001b[39mplugin, name\u001b[38;5;241m=\u001b[39mname)\n\u001b[1;32m    368\u001b[0m logger\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m        Start Nanny at: \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maddress)\n\u001b[0;32m--> 369\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39minstantiate()\n\u001b[1;32m    371\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response \u001b[38;5;241m!=\u001b[39m Status\u001b[38;5;241m.\u001b[39mrunning:\n\u001b[1;32m    372\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose(reason\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnanny-start-failed\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py:443\u001b[0m, in \u001b[0;36mNanny.instantiate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    441\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    442\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 443\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprocess\u001b[38;5;241m.\u001b[39mstart()\n\u001b[1;32m    444\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m    445\u001b[0m         logger\u001b[38;5;241m.\u001b[39merror(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFailed to start process\u001b[39m\u001b[38;5;124m\"\u001b[39m, exc_info\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py:720\u001b[0m, in \u001b[0;36mWorkerProcess.start\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    718\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstatus \u001b[38;5;241m=\u001b[39m Status\u001b[38;5;241m.\u001b[39mfailed\n\u001b[1;32m    719\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 720\u001b[0m     msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_wait_until_connected(uid)\n\u001b[1;32m    721\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m    722\u001b[0m     \u001b[38;5;66;03m# NOTE: doesn't wait for process to terminate, just for terminate signal to be sent\u001b[39;00m\n\u001b[1;32m    723\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprocess\u001b[38;5;241m.\u001b[39mterminate()\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py:868\u001b[0m, in \u001b[0;36mWorkerProcess._wait_until_connected\u001b[0;34m(self, uid)\u001b[0m\n\u001b[1;32m    865\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[1;32m    867\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexception\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m msg:\n\u001b[0;32m--> 868\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m msg[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexception\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    869\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    870\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m msg\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/nanny.py:919\u001b[0m, in \u001b[0;36mrun\u001b[0;34m()\u001b[0m\n\u001b[1;32m    917\u001b[0m failure_type: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minitialize\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    918\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 919\u001b[0m     worker \u001b[38;5;241m=\u001b[39m worker_factory()\n\u001b[1;32m    920\u001b[0m     failure_type \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstart\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    921\u001b[0m     thread \u001b[38;5;241m=\u001b[39m threading\u001b[38;5;241m.\u001b[39mThread(\n\u001b[1;32m    922\u001b[0m         target\u001b[38;5;241m=\u001b[39mfunctools\u001b[38;5;241m.\u001b[39mpartial(\n\u001b[1;32m    923\u001b[0m             watch_stop_q,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    928\u001b[0m         daemon\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    929\u001b[0m     )\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/distributed/worker.py:767\u001b[0m, in \u001b[0;36m__init__\u001b[0;34m()\u001b[0m\n\u001b[1;32m    727\u001b[0m handlers \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    728\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgather\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgather,\n\u001b[1;32m    729\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    751\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mget_story\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_story,\n\u001b[1;32m    752\u001b[0m }\n\u001b[1;32m    754\u001b[0m stream_handlers \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    755\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclose\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose,\n\u001b[1;32m    756\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcancel-compute\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_handle_remote_stimulus(CancelComputeEvent),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    764\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mremove-worker\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_handle_remove_worker,\n\u001b[1;32m    765\u001b[0m }\n\u001b[0;32m--> 767\u001b[0m ServerNode\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__init__\u001b[39m(\n\u001b[1;32m    768\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    769\u001b[0m     handlers\u001b[38;5;241m=\u001b[39mhandlers,\n\u001b[1;32m    770\u001b[0m     stream_handlers\u001b[38;5;241m=\u001b[39mstream_handlers,\n\u001b[1;32m    771\u001b[0m     connection_args\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconnection_args,\n\u001b[1;32m    772\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    773\u001b[0m )\n\u001b[1;32m    774\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmemory_manager \u001b[38;5;241m=\u001b[39m WorkerMemoryManager(\n\u001b[1;32m    775\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    776\u001b[0m     data\u001b[38;5;241m=\u001b[39mdata,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    781\u001b[0m     memory_pause_fraction\u001b[38;5;241m=\u001b[39mmemory_pause_fraction,\n\u001b[1;32m    782\u001b[0m )\n\u001b[1;32m    784\u001b[0m transfer_incoming_bytes_limit \u001b[38;5;241m=\u001b[39m math\u001b[38;5;241m.\u001b[39minf\n",
      "\u001b[0;31mTypeError\u001b[0m: __init__() got an unexpected keyword argument 'project'"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# get global alignment channels\n",
    "fix = fix_zarr['/c3/s3'][...]\n",
    "mov = mov_zarr['/c3/s3'][...][:, ::-1, ::-1]  # moving image data is 180 degrees rotates about z axis\n",
    "\n",
    "# define alignment steps\n",
    "ransac_kwargs = {\n",
    "    'blob_sizes':[4, 16],\n",
    "    'cc_radius':16,\n",
    "    'match_threshold':0.4,\n",
    "    'nspots':4000,\n",
    "}\n",
    "\n",
    "affine_kwargs = {\n",
    "    'alignment_spacing':2.0,\n",
    "    'shrink_factors':(2,),\n",
    "    'smooth_sigmas':(4.,),\n",
    "    'optimizer_args':{\n",
    "        'learningRate':0.25,\n",
    "        'minStep':0.,\n",
    "        'numberOfIterations':400,\n",
    "    },\n",
    "    'metric':'C',\n",
    "}\n",
    "\n",
    "steps = [('ransac', ransac_kwargs,),\n",
    "         ('affine', affine_kwargs,),]\n",
    "\n",
    "blocksize = [256, 256, 256]\n",
    "\n",
    "cluster_kwargs = {\n",
    "    'project':'scicompsoft',\n",
    "    'ncpus':4,\n",
    "    'threads':1,\n",
    "    'min_workers':60,\n",
    "    'max_workers':60,\n",
    "}\n",
    "\n",
    "# align\n",
    "deform = distributed_piecewise_alignment_pipeline(\n",
    "    fix, mov,\n",
    "    fix_spacing_s3,\n",
    "    mov_spacing_s3,\n",
    "    steps,\n",
    "    blocksize,\n",
    "    static_transform_list=[affine,],\n",
    "    cluster_kwargs=cluster_kwargs,\n",
    ")\n",
    "\n",
    "# apply affine only\n",
    "deform_aligned = apply_transform(\n",
    "    fix, mov,\n",
    "    fix_spacing_s3, mov_spacing_s3,\n",
    "    transform_list=[affine, deform],\n",
    ")\n",
    "\n",
    "# write results\n",
    "nrrd.write(outdir+'/deform.nrrd', deform, compression_level=2)\n",
    "nrrd.write(outdir+'/deformed.nrrd', deform_aligned.transpose(2,1,0), compression_level=2)\n",
    "\n",
    "# load precomputed results\n",
    "deform, _ = nrrd.read(outdir+'/deform.nrrd')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43789f30-7292-44b0-9a08-ee5e5617c1fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4408947d-d01d-4b0e-b9f9-5cb3a4dc6d2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1345b907-0943-4918-8bd8-084e9e66186b",
   "metadata": {},
   "source": [
    "### even more local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b87819f6-c082-421b-9538-b990432ef2f2",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'distributed_piecewise_alignment_pipeline' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "File \u001b[0;32m<timed exec>:37\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'distributed_piecewise_alignment_pipeline' is not defined"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# get global alignment channels\n",
    "fix = fix_zarr['/c3/s3'][...]\n",
    "mov = mov_zarr['/c3/s3'][...][:, ::-1, ::-1]  # moving image data is 180 degrees rotates about z axis\n",
    "\n",
    "deform_kwargs = {\n",
    "    'alignment_spacing':2.0,\n",
    "    'shrink_factors':(2,),\n",
    "    'smooth_sigmas':(4.,),\n",
    "    'control_point_spacing':200.0,\n",
    "    'control_point_levels':(1,),\n",
    "    'optimizer_args':{\n",
    "        'learningRate':2.5,\n",
    "        'minStep':0.07,\n",
    "        'numberOfIterations':75,\n",
    "    },\n",
    "    # 'metric':'C',  # correlation required too much memory, but the default MI worked well\n",
    "}\n",
    "\n",
    "steps = [('deform', deform_kwargs,),]\n",
    "         \n",
    "blocksize = [256, 256, 256]\n",
    "\n",
    "cluster_kwargs = {\n",
    "    'project':'scicompsoft',\n",
    "    'ncpus':3,\n",
    "    'threads':1,\n",
    "    'min_workers':60,\n",
    "    'max_workers':60,\n",
    "    'config':{\n",
    "        'distributed.worker.memory.target':0.9,\n",
    "        'distributed.worker.memory.spill':0.9,\n",
    "        'distributed.worker.memory.pause':0.9,\n",
    "    },\n",
    "}\n",
    "\n",
    "# align\n",
    "deform2 = distributed_piecewise_alignment_pipeline(\n",
    "    fix, mov,\n",
    "    fix_spacing_s3,\n",
    "    mov_spacing_s3,\n",
    "    steps,\n",
    "    blocksize,\n",
    "    static_transform_list=[affine, deform],\n",
    "    cluster_kwargs=cluster_kwargs,\n",
    ")\n",
    "\n",
    "# apply affine only\n",
    "deform2_aligned = apply_transform(\n",
    "    fix, mov,\n",
    "    fix_spacing_s3, mov_spacing_s3,\n",
    "    transform_list=[affine, deform, deform2],\n",
    ")\n",
    "         \n",
    "# write results\n",
    "nrrd.write(outdir+'/deform2.nrrd', deform2, compression_level=2)\n",
    "nrrd.write(outdir+'/deformed2.nrrd', deform2_aligned.transpose(2,1,0), compression_level=2)\n",
    "\n",
    "# load precomputed results\n",
    "deform2, _ = nrrd.read(outdir+'/deform2.nrrd')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbc6ea26-48f0-4fed-a2ac-52971a28a6b7",
   "metadata": {},
   "source": [
    "## Histogram equalize data to visualize alignment result better\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d61e9894-01f4-434d-a4fe-304bc80fb229",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/u/home/f/f7xiesnm/project-zipursky/easifish/results/test_bigstream2/fix_s3.nrrd'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m<timed exec>:4\u001b[0m\n",
      "File \u001b[0;32m~/.conda/envs/napari/lib/python3.9/site-packages/nrrd/reader.py:506\u001b[0m, in \u001b[0;36mread\u001b[0;34m(filename, custom_field_map, index_order)\u001b[0m\n\u001b[1;32m    474\u001b[0m \u001b[38;5;124;03m\"\"\"Read a NRRD file and return the header and data\u001b[39;00m\n\u001b[1;32m    475\u001b[0m \n\u001b[1;32m    476\u001b[0m \u001b[38;5;124;03mSee :ref:`user-guide:Reading NRRD files` for more information on reading NRRD files.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    502\u001b[0m \u001b[38;5;124;03m:meth:`write`, :meth:`read_header`, :meth:`read_data`\u001b[39;00m\n\u001b[1;32m    503\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    505\u001b[0m \u001b[38;5;124;03m\"\"\"Read a NRRD file and return a tuple (data, header).\"\"\"\u001b[39;00m\n\u001b[0;32m--> 506\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrb\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m fh:\n\u001b[1;32m    507\u001b[0m     header \u001b[38;5;241m=\u001b[39m read_header(fh, custom_field_map)\n\u001b[1;32m    508\u001b[0m     data \u001b[38;5;241m=\u001b[39m read_data(header, fh, filename, index_order)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/u/home/f/f7xiesnm/project-zipursky/easifish/results/test_bigstream2/fix_s3.nrrd'"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "from skimage.exposure import equalize_adapthist, equalize_hist\n",
    "from scipy.ndimage import gaussian_filter, grey_opening\n",
    "\n",
    "fix_s3, _ = nrrd.read(outdir+'/fix_s3.nrrd')\n",
    "deformed, _ = nrrd.read(outdir+'/deformed.nrrd')\n",
    "deformed2, _ = nrrd.read(outdir+'/deformed2.nrrd')\n",
    "\n",
    "fix_s3_corrected = equalize_hist(fix_s3, mask=(fix_s3 > 90))\n",
    "fix_s3_corrected = np.round(fix_s3_corrected * 64000).astype(np.uint16)\n",
    "\n",
    "deformed_corrected = equalize_hist(deformed, mask=(deformed > 90))\n",
    "deformed_corrected = np.round(deformed_corrected * 64000).astype(np.uint16)\n",
    "\n",
    "deformed2_corrected = equalize_hist(deformed2, mask=(deformed2 > 90))\n",
    "deformed2_corrected = np.round(deformed2_corrected * 64000).astype(np.uint16)\n",
    "\n",
    "nrrd.write(outdir+'/fix_s3_corrected.nrrd', fix_s3_corrected, compression_level=2)\n",
    "nrrd.write(outdir+'/deformed_corrected.nrrd', deformed_corrected, compression_level=2)\n",
    "nrrd.write(outdir+'/deformed2_corrected.nrrd', deformed2_corrected, compression_level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "468b7b52-f4bf-4ed5-bca2-239825252fe0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "napari",
   "language": "python",
   "name": "napari"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
